faiss_vector_cache.go

```diff
@@ -57,21 +64,65 @@ func (vc *vecIndexCache) Clear() {
 	vc.m.Unlock()
 }
 
-func (vc *vecIndexCache) loadVectorIndex(fieldID uint16,
-	indexBytes []byte) (vecIndex *faiss.IndexImpl, err error) {
-	cachedIndex, present := vc.isIndexCached(fieldID)
+func (vc *vecIndexCache) loadFromCache(fieldID uint16, mem []byte,
+	except *roaring.Bitmap) (index *faiss.IndexImpl,
+	vecDocIDMap map[int64]uint32, vecExcluded []int64, err error) {
+	entry, present := vc.checkEntry(fieldID)
+
 	if present {
-		vecIndex = cachedIndex
-		vc.incHit(fieldID)
+		index, vecDocIDMap, vecExcluded = entry.load(except)
 	} else {
-		vecIndex, err = vc.createAndCacheVectorIndex(fieldID, indexBytes)
+		index, vecDocIDMap, vecExcluded, err =
+			vc.createAndCacheEntry(fieldID, mem, except)
 	}
+
 	vc.addRef(fieldID)
-	return vecIndex, err
+	return index, vecDocIDMap, vecExcluded, err
 }
 
-func (vc *vecIndexCache) createAndCacheVectorIndex(fieldID uint16,
-	indexBytes []byte) (*faiss.IndexImpl, error) {
+func (vc *cacheEntry) load(except *roaring.Bitmap) (*faiss.IndexImpl,
+	map[int64]uint32, []int64) {
+
+	vc.m.RLock()
+	vecIndex := vc.index
+	vecDocIDMap := vc.vecDocIDMap
+	docIDExcluded := vc.docIDExcluded
+	vecExcluded := vc.vecExcluded
+	vc.m.RUnlock()
+
+	if except != nil {
+		newExcluded := false
+		vc.m.Lock()
+		it := except.Iterator()
+		for it.HasNext() {
```

Comment:
I don't think this is a good idea - we're trading compute for resource usage.
The least you should be doing is holding a copy of except and do ..
```
if cacheExcept.Intersects(except) {
``` 
.. to determine if the except bitmap has changed or not.

But even this is probably not worth the trade off of having to determine vecExcluded again in the event except has diverged.